plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableRedirectionsProvider.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hive;
+
+import io.trino.plugin.hive.metastore.SemiTransactionalHiveMetastore;
+import io.trino.spi.connector.CatalogSchemaTableName;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.connector.SchemaTableName;
+
+import java.util.Optional;
+
+public interface HiveTableRedirectionsProvider
+{
+    HiveTableRedirectionsProvider NO_REDIRECTS = (session, tableName, metastore) -> Optional.empty();
+
+    Optional<CatalogSchemaTableName> redirectTable(ConnectorSession session, SchemaTableName tableName, SemiTransactionalHiveMetastore metastore);
```

Comment:
not sure if we need to expose the full `SemiTransactionalHiveMetastore` here as it can also do updates which is too wide for a "redirect" API implementation. would it be feasible to just expose the "getters"? 

On the other hand, if this means we make the signature overly complex and hard to evolve, the current approach seems reasonable.

Reply:
I extracted some common logic around call to `RedirectionsProvider` and changed the interface to take `Table` object instead `metastore` instance (suggested by @findepi).
It may look bad as now we will (almost) always make the call to `metastore.getTable` to get the `Table` objects when `redirectTable` is called. 
But it should not matter from performance perspective as metastore calls are cached within the session, and we would get the call `metastore.getTable` anyway; just maybe a bit later. Now the latter call will come from cache.